File: MPI_Code/C/reduce2.c
#include <stdio.h>
#include <mpi.h>
int main()
{
  int nrank,i;
  double A[3];
  struct ddt{
    double val;
    int rank;
  };
  struct ddt in[3],out[3];
  MPI_Init(NULL,NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&nrank);
  if(nrank==0) A[0]=9, A[1]=5, A[2]=2;
  if(nrank==1) A[0]=8, A[1]=6, A[2]=1;
  if(nrank==2) A[0]=7, A[1]=4, A[2]=3;
  for(i=0;i<3;i++) in[i].val=A[i],  in[i].rank=nrank;
  MPI_Reduce(in,out,3,MPI_DOUBLE_INT,MPI_MAXLOC,0,MPI_COMM_WORLD);
  if(nrank==0){
    printf("max val : %.2f  %.2f  %2.f\n",out[0].val, out[1].val, out[2].val);
    printf("max rank: %d  %d  %d\n",out[0].rank, out[1].rank, out[2].rank);
  }
  MPI_Finalize();
  return 0;
}

File: example_code/mpi/MPI_Code/Fortran/reduce2.f90
PROGRAM reduce2
USE mpi_f08
IMPLICIT NONE
INTEGER::nrank,ista,iend,i,cnt=0
REAL::A(3)=0
TYPE DDT
  REAL::val,rank
END TYPE DDT
TYPE(DDT)::in(3),out(3)

CALL MPI_Init
CALL MPI_Comm_rank(MPI_COMM_WORLD,nrank)
IF(nrank==0) THEN
  A(1)=9;   A(2)=5;   A(3)=2
ELSEIF(nrank==1)THEN
  A(1)=8;   A(2)=6;   A(3)=1
ELSEIF(nrank==2)THEN
  A(1)=7;   A(2)=4;   A(3)=3
ENDIF
  
DO i=1,3
  in(i)%val=a(i)
  in(i)%rank=nrank
END DO
CALL MPI_Reduce(in,out,3,MPI_2REAL,MPI_MAXLOC, 0, MPI_COMM_WORLD)
IF(nrank==0)THEN
  PRINT*,'max val',out%val 
  PRINT*,'max rank',out%rank
ENDIF

CALL MPI_Finalize
END PROGRAM reduce2
